cmd/zfs/zfs_main.c

```diff
@@ -7961,17 +7962,31 @@ main(int argc, char **argv)
 
 	libzfs_print_on_error(g_zfs, B_TRUE);
 
+	/*
+	 * Many commands modify input strings for string parsing reasons.
+	 * We create a copy to protect the original argv.
+	 *
+	 * Under Linux the parameter maximum length is a fraction of the stack
+	 * size so this allocation should never overflow the stack.
+	 */
+	newargv = alloca((argc + 1) * sizeof (newargv[0]));
+	for (i = 0; i < argc; i++) {
+		newargv[i] = alloca(strlen(argv[i] + 1));
```

Comment:
Hi there! Thanks for working on this change.

Maybe we could change this:
```
for (i = 0; i < argc; i++) {
	newargv[i] = alloca(strlen(argv[i] + 1));
	strcpy(newargv[i], argv[i]);
}
```
to this:
```
for (i = 0; i < argc; i++) {
	newargv[i] = strdupa(argv[i]);
}
```
(in both cases where it is used)

To be honest though, I'm not sure how to feel about `alloca()` being
used as it is both platform and compiler dependent. Do we actually
get any wins from using it?

Reply:
Stack allocations are extremely fast don't have to be `free()`'d at the end of a program since they are effectively freed on return from `main()`. The alternative with regular `strdup()` is to call `free()` afterwards just for the sake of cleanliness. Not that big a deal but we have an alternative that, at least in the context of Linux, works equally as well.

Though platform and compiler dependent, it is well supported on gcc and the Linux man page implies it will be just fine on BSD as well. `strdupa` is noted as strictly a `gcc` thing whereas it seems `alloca` has a somewhat wider range of support. Heck I see an equivalent on MSDN. If `alloca` is not practical I'd go back to normal `strdup/free`.

Reply:
I have to admit I'm a bit skittish about using `alloca()` too.  I agree that it _should_ be portable, and it _shouldn't_  cause any issues.  But I'd prefer to be conservative and avoid using it unless there's a meaningful win to be had.  Let's follow the wise advice in the man page.

The alloca() function is machine- and compiler-dependent.  For  certain
applications,  its  use  can  improve efficiency compared to the use of
malloc(3) plus free(3).  In certain cases, it can also simplify  memory
deallocation  in  applications  that  use  longjmp(3) or siglongjmp(3).
__Otherwise, its use is discouraged.__

Plus this would be the only use of `alloca(3)` in the entire ZFS code base.  Why add another dependency unless we have too.